src/test/java/org/inferred/freebuilder/processor/util/testing/TempJavaFileManager.java

```diff
@@ -81,58 +126,105 @@ public boolean handleOption(String current, Iterator<String> remaining) {
+        key(location, className, kind),
+        k -> new InMemoryJavaFile(qualifiedName(className), kind));
+  }
+
+  private static QualifiedName qualifiedName(String className) {
+    int lastPeriod = className.lastIndexOf('.');
+    String packageName = lastPeriod >= 0 ? className.substring(0, lastPeriod) : "";
+    String remainder = className.substring(lastPeriod + 1);
+    List<String> simpleNames = Splitter.on('$').splitToList(remainder);
+    return QualifiedName.of(
+        packageName,
+        simpleNames.get(0),
+        simpleNames.subList(1, simpleNames.size()).toArray(new String[simpleNames.size() - 1]));
   }
 
   @Override
   public FileObject getFileForInput(Location location, String packageName, String relativeName)
       throws IOException {
-    return delegate.getFileForInput(location, packageName, relativeName);
+    if (LOCATIONS.contains(location)) {
+      Kind kind = kindFromExtension(relativeName);
+      Map.Entry<Location, String> key = key(location, packageName, relativeName);
+      return (KINDS.contains(kind) ? javaFiles : otherFiles).get(key);
+    } else {
+      return delegate.getFileForInput(location, packageName, relativeName);
+    }
+  }
+
+  private static Map.Entry<Location, String> key(
+      Location location, String packageName, String relativeName) {
+    return immutableEntry(location, path(packageName, relativeName));
+  }
+
+  private static String path(String packageName, String relativeName) {
+    return packageName.replace('.', '/')
+        + (packageName.isEmpty() ? "" : "/")
+        + relativeName;
+  }
+
+  private static Kind kindFromExtension(String relativeName) {
+    return Arrays.stream(Kind.values())
+        .filter(k -> relativeName.endsWith(k.extension) && k != OTHER)
+        .findAny()
+        .orElse(OTHER);
   }
 
   @Override
   public FileObject getFileForOutput(
-      Location location, String packageName, String relativeName, FileObject sibling)
-      throws IOException {
-    return delegate.getFileForOutput(location, packageName, relativeName, sibling);
+      Location location, String packageName, String relativeName, FileObject sibling) {
+    checkArgument(LOCATIONS.contains(location),
+        "Unsupported location %s (must be one of %s)", location, LOCATIONS);
+    Kind kind = kindFromExtension(relativeName);
+    Map.Entry<Location, String> key = key(location, packageName, relativeName);
+    if (KINDS.contains(kind)) {
+      return javaFiles.computeIfAbsent(key, k -> {
+        QualifiedName qualifiedName = qualifiedName(relativeName
+            .substring(0, relativeName.length() - kind.extension.length())
+            .replace('/', '.'));
+        return new InMemoryJavaFile(qualifiedName, kind);
+      });
+    } else {
+      return otherFiles.computeIfAbsent(key, k -> new InMemoryFile(k.getValue()));
+    }
   }
 
   @Override
-  public void flush() throws IOException {
-    delegate.flush();
-  }
+  public void flush() {}
```

Comment:
We're not delegating for `getJavaFileForOutput`, so I don't think there's a legitimate reason to delegate `flush`.